### PR TITLE
refactor: remove legacy map; add rule for vec'zd map

### LINKED ISSUE #4739: Add a way to assert about vectorized operations
Currently we don't have a way to write an end to end (flux acceptance) test which can make assertions about whether or not something in the query was actually vectorized.

Not exactly sure what the good solution for this is, but the closest thing I can think of available today is the API we have for asserting about pushdowns in the plan.

> related, but not necessarily on the critical path for this, but it would be interesting to be able to see information about vectorized operations from the profiler.

NOTE: We think the best way to assert about verctorized operations is to leverage the planner rules. This means we want to create/split the map transformation into two, one for row based and one for vectorized operations. Then the planner rule chooses between them when possible. This way enabling/disabling the planner rule controls vectorization. Additionally we can assert about the application of planner rules in tests already.

DOD:

- Remove old (i.e. no chunked implementation of map) this code path is currently dead in production because of the feature flag
- Split new (i.e. chunked implementation) of map transformation into two transformations, row and vectorized version.
- Create planner rule that picks the vectorized map transformation when the vectorized code is available
- Add/Update acceptance tests for each vector operation (i.e. logical, comparison, etc) that show correctness of the vectorized version asserting the planner rule was applied.

### PR DESCRIPTION
**Pull request by wolffcm**
Fixes #4739.

This PR removes the `vectorizedMap` feature flag, which has been on for a few months in production. Note that there is a deprecated parameter called `mergeKey` that was honored in the legacy implementation of `map`. It is going away altogether now, so there is some behavior change and I have updated the related Go unit tests.

In addition I added a planner rule that will enable the vectorized function to be executed when one exists. This way we can use `expect.planner` to ensure that `vectorizeMapRule` fires, and `vectorizedMap` will appear as a plan node in the graphviz rendering of a plan.

### Checklist

Dear Author :wave:, the following checks should be completed (or explicitly dismissed) before merging.

- [x] ✏️ Write a PR description, regardless of triviality, to include the _value_ of this PR
- [x] 🔗 Reference related issues
- [x] 🏃 Test cases are included to exercise the new code
- [x] 🧪 If **new packages** are being introduced to stdlib, link to Working Group discussion notes and ensure it lands under `experimental/` (N/A)
- [x] 📖 If **language features** are changing, ensure `docs/Spec.md` has been updated (N/A)

Dear Reviewer(s) :wave:, you are responsible (among others) for ensuring the completeness and quality of the above before approval.

### CODE REVIEW

File: stdlib/universe/map.go [hunk 1]
```diff
@@ -97,159 +105,264 @@ func createMapTransformation(id execute.DatasetID, mode execute.AccumulationMode
...
-	fn := execute.NewRowMapFn(spec.Fn.Fn, compiler.ToScope(spec.Fn.Scope))
-	return &mapTransformation{
-		d:        d,
-		cache:    cache,
-		fn:       fn,
-		ctx:      ctx,
-		mergeKey: spec.MergeKey,
-	}, nil
-}
+func (m *mapTransformation) Process(
```
Comment (wolffcm) [inline stdlib/universe/map.go:129]: This looks like a big change, but all of this code here was copied from `map2.go`. I did not need to alter it.

File: stdlib/universe/map_vectorized.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+	tr := &mapTransformation{
+		ctx: a.Context(),
+		fn: &mapVectorFunc{
+			fn: execute.NewVectorMapFn(s.Fn.Fn, compiler.ToScope(s.Fn.Scope)),
+		},
+	}
+	return execute.NewGroupTransformation(id, tr, a.Allocator())
+}
+
+type mapVectorFunc struct {
```
Comment (wolffcm) [inline stdlib/universe/map_vectorized.go:90]: Somehow Git lost track of it, but I meant to just change the name `map2.go` to `map_vectorized.go`. All of the code for the rest of the file was in `map2.go`, and I didn't need to change it.

File: stdlib/universe/map.go (hunk 1 above)
Comment (Marwes) [reply stdlib/universe/map.go:129]: Nit: If you delete the previous file first, then do `git mv` you can probably preserve the history better (might need to delete in one commit, move in the next).

**Review by Marwes — COMMENTED** [review]
I assume that any breakages from removing `mergeKey` has already been discussed and deemed to be ok? LGTM otherwise

**Comment by wolffcm** [comment]
> I assume that any breakages from removing mergeKey has already been discussed and deemed to be ok? LGTM otherwise

Yes, per our discussion this morning, `mergeKey` isn't going away as part of this PR, but I'll do it as a separate one.

File: stdlib/universe/map.go (hunk 1 above)
Comment (wolffcm) [reply stdlib/universe/map.go:129]: Ah, I will definitely do that next time.

**Review by Marwes — APPROVED** [review]

**Comment by wolffcm** [comment]
@theodiefenthal Hi Theo, there was an issue with `map` related to an optimization for if/then/else. We have since turned this optimization off in Cloud. Can you check and see if the issue still exists?

**Comment by onelson** [comment]
We'll hopefully know more soon, but my thinking is, to echo @wolffcm, the issue has more to do with the implementation for vectorized conditional expressions themselves, not the manner in which we select.

Investigation will likely happen over on https://github.com/influxdata/flux/issues/5155

**Comment by theodiefenthal** [comment]
Hi @wolffcm , the issue now disappeared for me, so turning it off in the cloud helped. Thank you. And my bad for adding this comment to the wrong ticket. I'll follow up on reading #5155.

**Pull request merged**